### PR TITLE
Feature: ed25519

### PR DESCRIPTION
**Pull request by metalurgical**
This PR enables ed25519 support.

Care must be taken when releasing this since it unfortunately contains numerous breaking changes.

### CODE REVIEW

File: packages/share-transfer/src/ShareTransferModule.ts [hunk 1]
```diff
@@ -51,7 +51,7 @@ class ShareTransferModule implements IModule {
 
     // This is needed to avoid MIM during share deletion.
     if (numberOfNewShares <= numberOfOldShares) {
-      const shareTransferStorePointer: ShareTransferStorePointer = { pointer: new BN(generatePrivate()) };
+      const shareTransferStorePointer: ShareTransferStorePointer = { pointer: new BN(generatePrivate(KeyType.secp256k1)) };
```
Comment (matthiasgeihs) [inline packages/share-transfer/src/ShareTransferModule.ts:54]: I assume `secp256k1` is intended here. Why does this functionality not depend on key type? Is this just used for encryption? Maybe create a function `generateEncryptionKey`?

File: packages/share-transfer/src/ShareTransferModule.ts [hunk 2]
```diff
@@ -87,11 +87,12 @@ class ShareTransferModule implements IModule {
...
     if (this.currentEncKey) throw ShareTransferError.requestExists(`${this.currentEncKey.toString("hex")}`);
-    this.currentEncKey = new BN(generatePrivate());
+    this.currentEncKey = new BN(generatePrivate(KeyType.secp256k1));
     const [newShareTransferStore, userIp] = await Promise.all([this.getShareTransferStore(), getClientIp()]);
-    const encPubKeyX = getPubKeyPoint(this.currentEncKey).x.toString("hex");
+    const encPubPoint = Point.fromPrivate(this.currentEncKey, KeyType.secp256k1);
+    const encPubKeyX = encPubPoint.x.toString("hex");
     newShareTransferStore[encPubKeyX] = new ShareRequest({
-      encPubKey: getPubKeyECC(this.currentEncKey),
+      encPubKey: Buffer.from(encPubPoint.toSEC1(), "hex"),
```
Comment (matthiasgeihs) [inline packages/share-transfer/src/ShareTransferModule.ts:95]: could use `getEncryptionPublicKey`?

File: packages/share-transfer/src/ShareTransferModule.ts [hunk 3]
```diff
@@ -72,7 +72,7 @@ class ShareTransferModule implements IModule {
     const rawShareTransferStorePointer = metadata.getGeneralStoreDomain(this.moduleName) as ShareTransferStorePointerArgs;
     let shareTransferStorePointer: ShareTransferStorePointer;
     if (!rawShareTransferStorePointer) {
-      shareTransferStorePointer = { pointer: new BN(generatePrivate()) };
+      shareTransferStorePointer = { pointer: new BN(generatePrivate(KeyType.secp256k1)) };
```
Comment (matthiasgeihs) [inline packages/share-transfer/src/ShareTransferModule.ts:75]: use dedicated function for generating encryption key?

**Review by matthiasgeihs — COMMENTED** [review]
Nice work!

Did a review focusing on cryptography related aspects. Mostly looks good. Mostly minor comments / suggestions / notes.

General note: I didn't see any instances where we generated `ed25519` 32-byte seed values. I was expecting this, for example, in case of importing keys. I don't know if this is done somewhere else (e.g., in frontend). In any case, if we want to be able to export an `ed25519` compatible private key at any later point in time, we need to make sure that we start by generating a 32-byte seed (not the scalar), and then store the encrypted seed somewhere for potential export later.

File: packages/common-types/src/base/BNUtils.ts [hunk 4]
```diff
@@ -1,24 +1,32 @@
...
 };
 
-export const getPubKeyEC = (bn: BN): curve.base.BasePoint => ecCurve.keyFromPrivate(bn.toString("hex", 64)).getPublic();
+export const getPubKeyEC = (bn: BN, keyType: KeyType): curve.base.BasePoint => {
+  return toPrivKeyEC(bn, keyType).getPublic();
+};
 
-export const getPubKeyECC = (bn: BN): Buffer => getPublic(toPrivKeyECC(bn));
+export const getPubKeyECC = (bn: BN, keyType: KeyType, compressed: boolean): Buffer => {
+  return Buffer.from(getPubKeyEC(bn, keyType).encode("array", compressed));
```
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/common-types/src/base/BNUtils.ts:26]: Yeah, your assistance will be appreciated here.

File: packages/common-types/src/base/Point.ts [hunk 5]
```diff
@@ -1,51 +1,49 @@
...
-  constructor(x: BNString, y: BNString) {
+  constructor(x: BNString, y: BNString, keyType: KeyType) {
     this.x = new BN(x, "hex");
     this.y = new BN(y, "hex");
+    this.keyType = keyType;
+  }
+
+  static fromJSON(value: StringifiedType): Point {
+    const { x, y, keyType } = value;
+    return new Point(x, y, keyType in KeyType ? keyType : KeyType.secp256k1);
```
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/common-types/src/base/Point.ts:20]: Backwards compatibility to be able to load existing data

File: packages/common-types/src/base/Polynomial.ts [hunk 6]
```diff
@@ -17,20 +15,29 @@ class Polynomial implements ISerializable {
...
+  constructor(polynomial: BN[], keyType: KeyType) {
     this.polynomial = polynomial;
+    this.keyType = keyType;
   }
 
   static fromJSON(value: StringifiedType): Polynomial {
-    const { polynomial } = value;
-    return new Polynomial(polynomial.map((x: string) => new BN(x, "hex")));
+    const { polynomial, keyType } = value;
+    const postKeyType = keyType in KeyType ? keyType : KeyType.secp256k1;
```
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/common-types/src/base/Polynomial.ts:27]: While I do agree, the problem is the use of `static` prevents using `this`.

File: packages/common-types/src/baseTypes/commonTypes.ts [hunk 7]
```diff
@@ -1,6 +1,24 @@
 import type { CustomAuthArgs } from "@toruslabs/customauth";
 import BN from "bn.js";
-import type { curve } from "elliptic";
+import { curve, ec as EllipticCurve } from "elliptic";
+
+export enum KeyType {
+  secp256k1 = "secp256k1",
+  ed25519 = "ed25519",
```
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/common-types/src/baseTypes/commonTypes.ts:7]: While I do agree, this has already been implemented elsewhere. We are conforming to this here.

File: packages/common-types/src/utils.ts [hunk 8]
```diff
@@ -1,20 +1,40 @@
...
+};
 
-// const privKeyBnToPubKeyECC = (bnPrivKey) => {
-//   return getPublic(privKeyBnToEcc(bnPrivKey));
-// };
+export function getEncryptionPrivateKey(privateKey: BN, keyType: KeyType): Buffer {
+  let priv = toPrivKeyEC(privateKey, keyType);
 
-export const ecCurve = new EC("secp256k1");
+  if (keyType === KeyType.ed25519) {
```
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/common-types/src/utils.ts:20]: I don't mind changing this to use SHA. Though I am pretty sure it's using keccak everywhere else, so we will have to check.

File: packages/common-types/src/utils.ts [hunk 9]
```diff
@@ -1,20 +1,40 @@
...
+  return toPrivKeyECC(priv.getPrivate(), KeyType.secp256k1);
+}
+
+export function getEncryptionPublicKey(privateKey: BN, keyType: KeyType): Buffer {
+  let priv = toPrivKeyEC(privateKey, keyType);
+
+  if (keyType === KeyType.ed25519) {
+    const secpCurve = keyTypeToCurve(KeyType.secp256k1);
+    priv = toPrivKeyEC(new BN(keccak512(toPrivKeyECC(privateKey, keyType))).umod(secpCurve.curve.n), KeyType.secp256k1);
+  }
```
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/common-types/src/utils.ts:34]: Mentioned function returns a buffer instead of a key. We could however make this block a small function in itself.

File: packages/storage-layer-torus/src/MockStorageLayer.ts [hunk 10]
```diff
@@ -57,43 +57,48 @@ class MockStorageLayer implements IStorageLayer {
    * @param input - data to post
    * @param privKey - If not provided, it will use service provider's share for encryption
    */
-  async setMetadata<T>(params: { input: T; serviceProvider?: IServiceProvider; privKey?: BN }): Promise<{ message: string }> {
-    const { serviceProvider, privKey, input } = params;
+  async setMetadata<T>(params: { input: T; serviceProvider?: IServiceProvider; privKey?: BN; keyType: KeyType }): Promise<{ message: string }> {
```
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/storage-layer-torus/src/MockStorageLayer.ts:60]: Yeah, we should add this.

File: packages/share-transfer/src/ShareTransferModule.ts (hunk 1 above)
Comment (metalurgical) [reply packages/share-transfer/src/ShareTransferModule.ts:54]: Hmm... this may be a bug here. Will re-check it.

Comment (metalurgical) [reply packages/share-transfer/src/ShareTransferModule.ts:54]: You are correct that secp256k1 is intended here though.

File: packages/common-types/src/baseTypes/commonTypes.ts [hunk 11]
```diff
@@ -1,6 +1,24 @@
...
+}
+const curveED25519 = new EllipticCurve("ed25519");
```
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/common-types/src/baseTypes/commonTypes.ts:9]: https://github.com/indutny/elliptic

The above link mentions that `ed25519` is a curve preset of the library and also makes the following statement:
`NOTE: That curve25519 could not be used for ECDSA, use ed25519 instead.`

The naming problem is part of the dependency...

File: packages/storage-layer-torus/src/MockStorageLayer.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/storage-layer-torus/src/MockStorageLayer.ts:60]: I think lets do this in a future PR. This one is quite large already.

File: packages/common-types/src/utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (metalurgical) [reply packages/common-types/src/utils.ts:21]: Comment was added.

**Comment by metalurgical** [comment]
@matthiasgeihs The seed you mentioned in your initial comment has been added. It was left for last due to the sheer number of tests that needed to pass with these changes.

**Comment by metalurgical** [comment]
Closes #240

File: packages/core/src/core.ts [hunk 12]
```diff
@@ -635,31 +654,35 @@ class ThresholdKey implements ITKey {
...
   } = {}): Promise<InitializeNewKeyResult> {
-    if (!importedKey) {
-      const tmpPriv = generatePrivate();
+    let seed: Uint8Array = new Uint8Array();
+    if (this.keyType === KeyType.secp256k1) {
+      const tmpPriv = importedKey || generatePrivate(this.keyType);
       this._setKey(new BN(tmpPriv));
     } else {
-      this._setKey(new BN(importedKey));
+      seed = importedKey ? new Uint8Array(importedKey.toBuffer()) : nacl.randomBytes(32);
```
Comment (matthiasgeihs) [inline packages/core/src/core.ts:662]: This sometimes fails. The imported key is a BN, but the seed needs to be a 32 byte array. Depending on how BN is generated, it may only be 31 bytes.

File: packages/core/src/core.ts [hunk 13]
```diff
@@ -635,31 +654,35 @@ class ThresholdKey implements ITKey {
     importedKey?: BN;
     delete1OutOf1?: boolean;
   } = {}): Promise<InitializeNewKeyResult> {
-    if (!importedKey) {
-      const tmpPriv = generatePrivate();
+    let seed: Uint8Array = new Uint8Array();
+    if (this.keyType === KeyType.secp256k1) {
+      const tmpPriv = importedKey || generatePrivate(this.keyType);
       this._setKey(new BN(tmpPriv));
     } else {
```
Comment (matthiasgeihs) [inline packages/core/src/core.ts:661]:
```suggestion
    } else if (this.keyType === KeyType.ed25519) {
      ...
    } else {
      // throw error
    }
```

File: packages/core/src/core.ts (hunk 12 above)
Comment (metalurgical) [reply packages/core/src/core.ts:662]: This would be a valid failure due to incorrect input then. No?

Comment (matthiasgeihs) [reply packages/core/src/core.ts]: The problem is that `BN.toBuffer` is not guaranteed to output 32 bytes but `nacl.sign.keyPair.fromSeed(seed);` on line 663 requires the seed to be 32 bytes. For example, consider the case where the 32-byte seed starts with a zero.

When I ran the tests locally, this was failing. In fact, the test (see below) was not even generating a 32 byte seed, but instead was generating a random scalar.

https://github.com/tkey/tkey/blob/15197375a8b792fa13b54066d6f0a3336690fdd0/packages/default/test/shared.js#L116

I think it's confusing that the imported key is a BN in the context of ed25519. An alternative would be to have the imported key be a byte array that either gets converted to BN (in case of secp256k1) or is consumed directly as a seed (in case of ed25519). What do you think?

Comment (matthiasgeihs) [reply packages/core/src/core.ts]: Seems to be resolved in https://github.com/tkey/tkey/pull/266/commits/15197375a8b792fa13b54066d6f0a3336690fdd0

File: .github/workflows/backward.yml [hunk 14]
```diff
@@ -4,11 +4,12 @@ name: Backward Compatibility Tests
...
+  create:
   # Triggers the workflow on push or pull request events but only for the master branch
-  push:
-    branches: [master]
-  pull_request:
-    branches: [master]
+  #push:
+  #  branches: [master]
+  #pull_request:
+  #  branches: [master]
```
Comment (matthiasgeihs) [inline .github/workflows/backward.yml:12]: Is this change intentional? Should it be kept like this? If yes, should remove outcommented lines (to keep it clean).

File: package.json [hunk 15]
```diff
@@ -80,5 +80,8 @@
   "engines": {
     "node": ">=18.x",
     "npm": ">=9.x"
+  },
+  "dependencies": {
+    "ethereum-cryptography": "^2.1.3"
```
Comment (matthiasgeihs) [inline package.json:85]: The root `package.json` shouldn't have dependencies. These should be place in the corresponding package's `package.json`.

File: .github/workflows/backward.yml (hunk 14 above)
Comment (himanshuchawla009) [reply .github/workflows/backward.yml:12]: yes, so currently backward compat tests are broken on master branch, we need a seperate PR to fix this.

File: packages/core/src/core.ts [hunk 16]
```diff
@@ -635,31 +654,37 @@ class ThresholdKey implements ITKey {
...
-      const tmpPriv = generatePrivate();
+    let seed: Uint8Array = new Uint8Array();
+    if (this.keyType === KeyType.secp256k1) {
+      const tmpPriv = importedKey || generatePrivate(this.keyType);
       this._setKey(new BN(tmpPriv));
+    } else if (this.keyType === KeyType.ed25519) {
+      seed = importedKey ? new Uint8Array(importedKey.toBuffer(undefined, 32)) : nacl.randomBytes(32);
+      const keyPair = nacl.sign.keyPair.fromSeed(seed);
+      const tempPriv = new BN(keyPair.secretKey.slice(0, 32)).umod(this.ecCurve.curve.n);
+      this._setKey(tempPriv);
```
Comment (matthiasgeihs) [inline packages/core/src/core.ts:665]: This looks odd.

`keyPair.secretKey.slice(0, 32) == seed`

Are we sure we want the `privKey` to be the seed? Shouldn't it be the derived scalar instead?

@metalurgical @himanshuchawla009

Comment (matthiasgeihs) [reply packages/core/src/core.ts:665]: Or it doesn't matter? But then why not just write `new BN(seed).umod(this.ecCurve.curve.n)`?

Also, should we make the `importedKey` of type `Buffer` or `Uint8Array` and require it to be 32 bytes?

Comment (himanshuchawla009) [reply packages/core/src/core.ts:665]: i have fixed all this in my PR, plz wait for 30 mins before reviewing, facing some typescript issues, which is blocking me to merge the PR

**Comment by metalurgical** [comment]
@himanshuchawla009 Please fix the tests and conflicts from merging #267

File: packages/core/src/core.ts [hunk 17]
```diff
@@ -682,7 +714,10 @@ class ThresholdKey implements ITKey {
     // acquireLock: false. Force push
     await this.addLocalMetadataTransitions({ input: [...authMetadatas, shareStore], privKey: [...sharesToPush, undefined] });
     if (delete1OutOf1) {
-      await this.addLocalMetadataTransitions({ input: [{ message: ONE_KEY_DELETE_NONCE }], privKey: [this.serviceProvider.postboxKey] });
+      await this.addLocalMetadataTransitions({
+        input: [{ message: ONE_KEY_DELETE_NONCE }],
+        privKey: [this.serviceProvider.postboxKey],
```
Comment (ieow) [inline packages/core/src/core.ts:719]: is it possible the postboxkey is ed25519?

Comment (metalurgical) [reply packages/core/src/core.ts:719]: Yes. For ed25519 the postbox key would also be ed25519.

Comment (metalurgical) [reply packages/core/src/core.ts:719]: It would also be the scalar value.

File: packages/share-transfer/src/ShareTransferModule.ts [hunk 18]
```diff
@@ -72,7 +76,7 @@ class ShareTransferModule implements IModule {
     const rawShareTransferStorePointer = metadata.getGeneralStoreDomain(this.moduleName) as ShareTransferStorePointerArgs;
     let shareTransferStorePointer: ShareTransferStorePointer;
     if (!rawShareTransferStorePointer) {
-      shareTransferStorePointer = { pointer: new BN(generatePrivate()) };
+      shareTransferStorePointer = { pointer: new BN(getEncryptionPrivateKey(generatePrivate(metadata.keyType), KeyType.secp256k1)) };
```
Comment (ieow) [inline packages/share-transfer/src/ShareTransferModule.ts:79]: why the pointer need to be typed?
pointer can be explicit to be a secp256k1 type.

Comment (ieow) [reply packages/share-transfer/src/ShareTransferModule.ts:79]: only the share that is being transfer need to be typed. 
other pointer, encryption can be done in secp256k1 only

**Comment by ieow** [comment]
Some question
we have keyType (field) for the following
- serviceProvider ( oauth/postbox key)
- tkey
- tss

I think we should not have mix of keyType for serviceProvider and Tkey and Tss 

as the postbox key is used in tkey, and it derive the encryption key from the tkey's keyType
the tss need the match the serviceProvider's keyType so that it could sign using same keyType shares

**Comment by matthiasgeihs** [comment]
> as the postbox key is used in tkey, and it derive the encryption key from the tkey's keyType the tss need the match the serviceProvider's keyType so that it could sign using same keyType shares

In my proposal #268, there is a special provider `TSSTorusServiceProvider` in order to resolve this conflict. This provider has a separate `tssKeyType` for anything related to TSS. (see [provider.ts](https://github.com/tkey/tkey/blob/e9d40f45c224954c60225435a4a76b116af0db18/packages/tss/src/provider.ts))

File: packages/core/src/core.ts (hunk 17 above)
Comment (himanshuchawla009) [reply packages/core/src/core.ts:719]: yes, that's correct. It ed25519 scalar

File: packages/share-transfer/src/ShareTransferModule.ts (hunk 18 above)
Comment (himanshuchawla009) [reply packages/share-transfer/src/ShareTransferModule.ts:79]: @metalurgical

**Comment by himanshuchawla009** [comment]
> Should we add check on the keyType for service Provider and the tkey?

wht type of check are you suggesting? Can you explain a bit more.

**Comment by himanshuchawla009** [comment]
> @himanshuchawla009 Please fix the tests and conflicts from merging #267

fixed

File: packages/share-transfer/src/ShareTransferModule.ts (hunk 18 above)
Comment (metalurgical) [reply packages/share-transfer/src/ShareTransferModule.ts:79]: @ieow  Oh, I just though it best to be kept consistent everywhere. This would still be be correct to do, no?

File: packages/core/src/core.ts [hunk 19]
```diff
@@ -1241,6 +1294,25 @@ class ThresholdKey implements ITKey {
...
+    if (!this.privKey) {
+      throw CoreError.privateKeyUnavailable();
+    }
+
+    if (this.keyType === KeyType.secp256k1) {
+      return this.privKey.toString("hex");
+    } else if (this.keyType === KeyType.ed25519) {
+      const result = this.metadata.getGeneralStoreDomain("ed25519Seed") as Record<string, EncryptedMessage>;
+      const seed = new BN(await this.decrypt(result.message));
+      return seed.toString("hex");
```
Comment (matthiasgeihs) [inline packages/core/src/core.ts:1311]: This should not be converted to BN. Instead, simply convert the buffer to hex.

File: .github/workflows/backward.yml (hunk 14 above)
Comment (metalurgical) [reply .github/workflows/backward.yml:12]: All these tests were fixed here? When were they broken?

Comment (metalurgical) [reply .github/workflows/backward.yml:12]: Nevermind, you are correct here.

**Review by ieow — APPROVED** [review]
lgtm
